### ISSUE TITLE
Unable to get kiali pod up running with demo profile

### ISSUE THREAD

**Issue by linsun** [istio/istio#11244]
grabbed latest 01/24's build, installed istio-demo.yaml.  Kiali pod stucks in container config error as below.

```
kiali-69d7f779c8-9tjcr                                         0/1     CreateContainerConfigError   0          9s
```

**Comment by jmazzitelli** [comment]
```
Warning Failed 4m12s (x72 over 19m) kubelet, 10.188.142.223 Error: secret "kiali" not found```
```

As [per the Istio docs](https://istio.io/docs/tasks/telemetry/kiali/#before-you-begin), you have to create the Kiali secret first. The Istio team asked that we NOT provide a default secret or default credentials in the helm charts for security purposes; the user must create one prior to installing Kiali.

**Comment by linsun** [comment]
I see, but that is an odd experience for quick install path as user will be looking at pods reach running before even getting to the kiali doc in telemetry section.

**Comment by clyang82** [comment]
Yes. that is not a good experience. I remembered there is the same issue raised before. but it is closed. 

since kiali v0.11, the authentication can be disabled. so we can consider to upgrade to v0.13 then kiali is not required to create secret anymore. just anonymous login allowed.  refer to https://issues.jboss.org/browse/KIALI-1958?_sscc=t

**Comment by sdake** [comment]
Istio should be secure by default.  I believe this should be true for cases of istio-demo.yaml as well.  I don't have any good solutions for setting credentials during a quick installation path.  I agree seeing an error is not intuitive.

Unfortunately, we can't pass the password via values.yaml (because `helm get values` can return the creds).  This is why we reference it via a secret.  Perhaps a better solution is for Kiali to startup and not report any particular error - but also not permit authorization until the Kiali secret is registered.

I believe an anonymous login for criitical mesh data is not a good solution either.

@jmazzitelli thoughts?

Cheers
-steve

**Comment by jmazzitelli** [comment]
As @sdake said, Istio standards dictate that we not provide a default, canned username/password, nor allow it to be unsecured by default (this is why we took out the canned "admin" username/password in the latest helm chart - we used to have it but removed it).

I don't see how Kiali can startup and not report an error (remember, it isn't Kiali reporting the error here - it's Kubernetes not being able to start up the pod because [the Kiali deployment's env vars can't be set due to the missing secret](https://github.com/istio/istio/blob/release-1.1/install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml#L46-L55)). I haven't looked at the Kubernetes docs on this feature - is it possible to have k8s not set the env var (or set it to empty string) but still start the pod if the declared secret does not exist? 

Another alternative would be to not require the secret to exist for deployment and not define those env vars at all, but have Kiali look up the secret at startup using the k8s API. If the secret is found missing, Kiali can serve up an error page. I can look into this if we feel this is a viable option. What do people think about that?

**Comment by jmazzitelli** [comment]
FYI: I think we can make these env vars based on secrets optional - and that would allow kiali to still start but present a more user-friendly error message. I read this at https://kubernetes.io/docs/concepts/configuration/secret/

```https://kubernetes.io/docs/concepts/configuration/secret/
Secrets must be created before they are consumed in pods as environment variables unless
they are marked as optional. References to Secrets that do not exist will prevent
the pod from starting.

References via secretKeyRef to keys that do not exist in a named Secret
will prevent the pod from starting.

Secrets used to populate environment variables via envFrom that have keys that are
considered invalid environment variable names will have those keys skipped. The
pod will be allowed to start. There will be an event whose reason is
InvalidVariableNames and the message will contain the list of invalid keys
that were skipped. The example shows a pod which refers to the default/mysecret
that contains 2 invalid keys, 1badkey and 2alsobad.
```

**Comment by linsun** [comment]
@jmazzitelli would it be reasonable to not provide default secret/credential in helm chart but provide it in the istio-demo.yaml when building that yaml file?  I think that will provide the best experience for users trying istio using the istio-demo.yaml one step install.

**Comment by jmazzitelli** [comment]
> would it be reasonable to not provide default secret/credential in helm chart but provide it in the istio-demo.yaml when building that yaml file

That would probably be the easiest solution if that would be allowed (having default credentials when installing using the demo.yaml). By "allowed" I mean is that OK with the Istio team to do this?

**Comment by linsun** [comment]
(I apologize I entered my comment without refreshing my browser so I missed all your comments :)

> That would probably be the easiest solution if that would be allowed (having default credentials when installing using the demo.yaml). By "allowed" I mean is that OK with the Istio team to do this?

Yes, I would be okay with that, demo profile is for demo purpose and not for production, people should know that.  It is targeted to provide the best demo experience for istio, else, I'd vote for not providing the istio-demo.yaml.

> I think we can make these env vars based on secrets optional - and that would allow kiali to still start but present a more user-friendly error message. 

I also like this option too...

**Comment by jmazzitelli** [comment]
@linsun Let me see if I can get helm to create a default secret and turn that on via a value from the demos values file. Just so I'm clear - we are talking about allowing something like this via helm:

`helm --set kiali.createDefaultSecret=true ...`

where that setting would be on by default in the [values-istio-demo.yaml](https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/values-istio-demo.yaml) - like so:

```
kiali:
  enabled: true
  createDefaultSecret: true
```

And somehow that ".Values.createDefaultSecret" would be used in the kiali helm subchart to create a default kiali secret with the default credentials of "admin" / "admin".

Am I understanding that correctly?

That would probably be the easiest thing to do, and would be good from a demo perspective.

For a production installation, I think we could do something with the envFrom feature. We could always have helm create an empty secret that we use with the envFrom feature (so the pod will always have a secret and thus will always be able to start). However, if the credentials env vars are missing from the secret, those env vars will be empty/not set - Kiali can detect this, and spit out a user-friendly error message on the login page telling the user what needs to be done. I wrote this up in the KIALI JIRA project for tracking: https://issues.jboss.org/browse/KIALI-2303

**Comment by jmazzitelli** [comment]
@linsun please review PR #11272 - that should install a secret when using the demo (or demo-auth) yaml. The default credentials will be username=admin and passphrase=admin.

**Comment by jmazzitelli** [comment]
OK, this issue with letting kiali start even with a missing secret is easier than I thought. I'll just add the "[optional](https://github.com/kubernetes/kubernetes/blob/e9e190c57c61a88a2ba96c4b0688b8eb3a8945a8/pkg/apis/core/types.go#L1770)" flag on the secret declaration in the Deployment. Kiali can then handle the empty env var to spit out a friendly error msg. This work will be done via https://issues.jboss.org/browse/KIALI-2303 .  We'll update the helm chart with this new optional flag when the new kiali is released that implements this change.

**Comment by linsun** [comment]
+1 @jmazzitelli thanks, agreed with your plan on the optional flag.  It'd be much easier to use and debug.

**Comment by jmazzitelli** [comment]
The work is done for the feature to allow kiali to start when the secret is missing, but have login failure message tell the user what the problem is. This is JIRA https://issues.jboss.org/browse/KIALI-2303

See PRs:
 * https://github.com/kiali/kiali/pull/811
 * https://github.com/kiali/kiali-ui/pull/964

This is in the process of being reviewed and then it will be merged, but this should make it into the next Kiali release.

**Comment by linsun** [comment]
Awesome, thank you @jmazzitelli

**Comment by sdake** [comment]
@jmazzitelli nice work!

**Comment by linsun** [comment]
hmm, pulled down latest 1.1 release - istio-release-1.1-20190212-09-1

still getting the same error.  @jmazzitelli perhaps the config isn't propogated to istio-demo.yaml?

```
  Normal   Scheduled    7m56s                   default-scheduler        Successfully assigned istio-system/kiali-69d7f779c8-rfvl6 to 10.188.142.194
  Warning  FailedMount  7m53s (x2 over 7m55s)   kubelet, 10.188.142.194  MountVolume.SetUp failed for volume "kiali-service-account-token-6zlg6" : couldn't propagate object cache: timed out waiting for the condition
  Warning  FailedMount  7m53s (x2 over 7m55s)   kubelet, 10.188.142.194  MountVolume.SetUp failed for volume "kiali-configuration" : couldn't propagate object cache: timed out waiting for the condition
  Normal   Pulling      7m51s                   kubelet, 10.188.142.194  pulling image "docker.io/kiali/kiali:v0.12"
  Normal   Pulled       7m36s                   kubelet, 10.188.142.194  Successfully pulled image "docker.io/kiali/kiali:v0.12"
  Normal   Pulled       5m55s (x9 over 7m36s)   kubelet, 10.188.142.194  Container image "docker.io/kiali/kiali:v0.12" already present on machine
  Warning  Failed       2m53s (x24 over 7m36s)  kubelet, 10.188.142.194  Error: secret "kiali" not found
```

**Comment by jmazzitelli** [comment]
@linsun this new feature (kiali secret installed via demo profile) was only introduced in the master branch via PR https://github.com/istio/istio/pull/11272/files

You'll see this is not in release-1.1. It was my understanding only high priority bug fixes and the like were getting in the relase-1.1.

Should I create a PR to cherry-pick that over into the release-1.1 branch?

**Comment by jmazzitelli** [comment]
@linsun I created a release-1.1 PR in case you want to get this in the 1.1 release. See https://github.com/istio/istio/pull/11715

**Comment by linsun** [comment]
@jmazzitelli thanks, sorry I had thought it was fixed in 1.1 already.  I'll try to see if we can get it in 1.1

**Comment by linsun** [comment]
this is fixed in 1.1

**Comment by Stormarov45** [comment]
I had just debugged an issue with the current demo profile for the current Istio version for x86 for Linux. 

I reference that here:  #https://github.com/istio/istio.io/issues/7572
